### PR TITLE
[Bounty]  Support Mamba 2.8B model

### PR DESCRIPTION
**Pull request by pollachisaravanan**
The PR will add Mamba 2.8B model,

1. Adapted the Mamba reference implementation from the TT-Metal repository(i.e https://github.com/tenstorrent/tt-metal/blob/main/models/demos/wormhole/mamba/reference/model.py)
2. Removed the [floating point conversion](https://github.com/tenstorrent/tt-metal/blob/a2c29371988a5902d6a67bc44416eb1d4dc4177d/models/demos/wormhole/mamba/reference/model.py#L295) and [specified dtype argument in zeros tensor construction](https://github.com/tenstorrent/tt-metal/blob/a2c29371988a5902d6a67bc44416eb1d4dc4177d/models/demos/wormhole/mamba/reference/model.py#L349) present in the model implementation which helps to run model in bfloat16 dtype.
3. Added mamba model test suite in the`tests/` directory path.
4. Lowered `torch.ops.aten.softplus.default` to `ttnn.softplus` and `torch.ops.aten.split_with_sizes.default` to `ttnn.split` and [ttnn.split accepts split_sizes as List/int](https://github.com/tenstorrent/tt-metal/blob/4deffa4c804c8c4376f7e108e1c20de9321ca2ef/ttnn/cpp/ttnn/operations/data_movement/split/split_pybind.cpp#L62C1-L62C62) so removed the condition in `ReplaceMoreTt`. 
5. Added operation tests suite for softplus and split ops.
6. The Mamba 2.8B model failing with `Out of Memory: Not enough space to allocate 167772160 B DRAM buffer across 12 banks, where each bank needs to store 13981696 B (assert.hpp:107)` error in compilation similar to [llama ](https://github.com/tenstorrent/pytorch2.0_ttnn/blob/main/tests/models/llama/test_llama.py). To resolve the issues, enabled the  EvictionPass which helps to compile and executed sucessfully for batch size of 1 and 32 with accuracy of 0.99 and all the atens ops are converted into ttnn ops 
7. Generated report by using the tools/collect_metrics.py file for checking the status of the model running e2e and profiling. For reference, I am attaching the report table here.

| Model                                                              | Status   |   Batch |   Compiled First Run (ms) |   Original Throughput (Inferences Per Second) |   Compiled Throughput (Inferences Per Second) |   Accuracy (%) | Torch Ops Before (Unique Ops)   | Torch Ops Remain (Unique Ops)   |   To/From Device Ops |
|:-------------------------------------------------------------------|:---------|--------:|--------------------------:|----------------------------------------------:|----------------------------------------------:|---------------:|:--------------------------------|:--------------------------------|---------------------:|
| [mamba-2.8b-batch_size-1](<docs/models/mamba-2.8b-batch_size-1>)   | ✅       |       1 |                    173293 |                                     0.0614344 |                                    0.00577056 |          99.98 | 48739 (23)                      | 0 (0)                           |                    0 |
| [mamba-2.8b-batch_size-32](<docs/models/mamba-2.8b-batch_size-32>) | ✅       |      32 |                    197954 |                                     0.890393  |                                    0.161654   |          99.97 | 48804 (24)                      | 0 (0)                           |                    0 |

 and also profiled using PyTorch built-in profiler and attaching the profiler results. 
[pytorch-stat.zip](https://github.com/user-attachments/files/21538046/pytorch-stat.zip)

8. Added pytest option for disabling the `LoadOncePass  `using `--load_params_once` and enbaling the `EvictionPass ` using` --run_eviction_opt` and PyTorch's built-in profiler using `--profile` in confest.py

**Mamba 2.8B Model with batch size of 1 and 32 logs**
[mamba_batch_size_1_and_32.log](https://github.com/user-attachments/files/21538017/mamba_batch_size_1_and_32.log)

**Split and Softplus Logs:**

[split.log](https://github.com/user-attachments/files/21299260/split.log)
[softplus.log](https://github.com/user-attachments/files/21299261/softplus.log)

### CODE REVIEW

**Comment by pollachisaravanan** [comment]
@ayerofieiev-tt @cguerreroTT  Could you please provide guidance on how we might proceed to support the full mamba 2.8B model with 64 layers?

**Comment by pollachisaravanan** [comment]
@ayerofieiev-tt @cguerreroTT I have tried compiling the Mamba 2.8B model with memory analysis pass(i.e run_mem_analysis = True in TorchTtnnOption). The Mamba 2.8B model compiled and executed successfully. 
Reference tests: [tests/memory/mamba/test_mamba.py](https://github.com/pollachisaravanan/pytorch2.0_ttnn/blob/2bacd2f3911fe61a1a2d970015fa09ed6e5781fb/tests/memory/mamba/test_mamba.py)

So I have added pytest option(i.e[ --run_mem_analysis](https://github.com/pollachisaravanan/pytorch2.0_ttnn/blob/2bacd2f3911fe61a1a2d970015fa09ed6e5781fb/tests/conftest.py#L55C1-L59C6)) which help to enable memory analysis for compile_and_run fixture and now the mamba model test suite(i.e [tests/models/mamba/test_mamba.py::test_mamba](https://github.com/pollachisaravanan/pytorch2.0_ttnn/blob/2bacd2f3911fe61a1a2d970015fa09ed6e5781fb/tests/models/mamba/test_mamba.py#L32)) compiled and executed successfully with accuracy of 0.99 and all the aten ops are converted to ttnn ops and I have generated report by using [tools/collect_metrics.py](https://github.com/tenstorrent/pytorch2.0_ttnn/blob/main/tools/collect_metrics.py)

File: tests/models/mamba/test_mamba.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+import pytest
+import torch
+from transformers import AutoTokenizer
+
+from tests.models.mamba.mamba_ssm.model import Mamba
+from tests.utils import ModelTester, repeat_inputs
+from loguru import logger
+
+torch._dynamo.config.suppress_errors = True
```
Comment (kevinwuTT) [inline tests/models/mamba/test_mamba.py:13]: Can you remove this line? We generally do not want to suppress any errors. Any error should be thoroughly investigated.

Comment (pollachisaravanan) [reply tests/models/mamba/test_mamba.py]: I have removed the `torch._dynamo.config.suppress_errors = True` and investigated the error's and shared the insight in this [bounty issue comment](https://github.com/tenstorrent/pytorch2.0_ttnn/issues/1048#issuecomment-3271875658).

**Comment by marty1885** [comment]
Hello @pollachisaravanan Could you resolve the conflicts? And I see the review has been stuck. Anything I can to make progress on our end?

**Comment by pollachisaravanan** [comment]
Hi @marty1885  — I have a question about supporting the Mamba 2.8B model with a batch size of 32. I posted the details in this bounty issue (https://github.com/tenstorrent/pytorch2.0_ttnn/issues/1048#issuecomment-3366378724) and asked @kevinwuTT for input, but I haven’t heard back yet. @kevinwuTT  — could you please share your thoughts on the comment when you have a moment? The reason I haven’t progressed is that I’m not seeing support for any models with batch sizes greater than 1. Given that, would it be acceptable to merge the current patch which enables Mamba 2.8B with batch size = 1 (consistent with other models in this repo), and then follow up with a separate change to add larger-batch support?

**Comment by ayerofieiev-tt** [comment]
@jmalone-tt , fyi

File: tests/models/mamba/mamba_ssm/model.py [hunk 2]
```diff
@@ -0,0 +1,382 @@
+# SPDX-FileCopyrightText: © 2025 Tenstorrent Inc.
```
[the remark this answers is not in the document]
Comment (pollachisaravanan) [reply tests/models/mamba/mamba_ssm/model.py]: In tt-metal, they also reimplemented the Mamba SSM model instead of loading the model similar to Hugging Face page show because the mamba-ssm package requires GPU/CUDA to install. To avoid those dependencies and ensure the model works in CPU-only and non-CUDA environments, we follow the same approach and provide a self-contained implementation.

File: tests/conftest.py [hunk 3]
```diff
@@ -52,7 +53,21 @@ def pytest_addoption(parser):
...
+        action="store_true",
+        help="Disable LoadParamsOnce optimization.",
+    )
+    parser.addoption(
+        "--run_eviction_opt", 
+        action="store_true", 
+        help="Enable the eviction optimization pass.",
+    )
+    parser.addoption(
+        "--profile",
```
[the remark this answers is not in the document]
Comment (pollachisaravanan) [reply tests/conftest.py]: Removed profiling option

File: tests/conftest.py [hunk 4]
```diff
@@ -52,7 +53,21 @@ def pytest_addoption(parser):
...
         help="Use native device integration for ttnn. Note: this is not supported with data parallel.",
     )
-
+    parser.addoption(
+        "--disable_load_params_once",
+        action="store_true",
+        help="Disable LoadParamsOnce optimization.",
+    )
+    parser.addoption(
+        "--run_eviction_opt",
```
[the remark this answers is not in the document]
Comment (pollachisaravanan) [reply tests/conftest.py]: The mamba 2.8B model can be compiled and executed either by enabling eviction pass(with additional memory pass fix introduced in this PR) or disabling load param once. I have created the pytest option to enable it(default: disabled). But I will add a commit to remove the option if needed we can revert it and  will retaining the memory pass fix.
